### PR TITLE
Introduce pre-commit and shellcheck

### PR DESCRIPTION
**Pull request by aledbf**
**Changes:**

- Bump Go version to 1.16.5
- Bump nvm.sh to v0.38.0
- Introduce pre-commit checks
- Run shellcheck in shell scripts

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,6 +1,19 @@
-version: 2.0
+version: 2.1
 
 jobs:
+  pre-commit:
+    docker:
+      - image: ubuntu:20.10
```
Comment (jankeromnes) [inline .circleci/config.yml:6]: Question: Why not an LTS?

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,13 @@
 
 A curated, chronologically ordered list of notable changes in [Gitpod's default workspace images](https://hub.docker.com/u/gitpod).
 
+## 2020-06-04
```
Comment (jankeromnes) [inline CHANGELOG.md:5]: Nit: It's 2021 already 😋

```suggestion
## 2021-06-04
```

File: full/nvm-lazy.sh [hunk 3]
```diff
@@ -1,18 +1,25 @@
...
 
-load_nvm () {
-    export NVM_DIR=~/.nvm
-    [ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"  # This loads nvm
-    [ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
-}
+    # Lazy-loading nvm + npm on node globals
+    load_nvm () {
+        # shellcheck disable=SC1091,SC1090
+        [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
```
Comment (jankeromnes) [inline full/nvm-lazy.sh:15]: Question: Is sourcing `bash_completion` no longer required?

File: full/nvm-lazy.sh [hunk 4]
```diff
@@ -1,18 +1,25 @@
...
-for cmd in "${NODE_GLOBALS[@]}"; do
-    eval "${cmd}(){ unset -f ${NODE_GLOBALS[@]}; load_nvm; ${cmd} \$@; }"
-done
+    # Making node global trigger the lazy loading
+    for cmd in "${NODE_GLOBALS[@]}"; do
+        # shellcheck disable=SC2128
+        eval "${cmd}(){ unset -f ${NODE_GLOBALS}; load_nvm; ${cmd} \$@ }"
+    done
+
+    export PATH="$PATH:$HOME/.yarn/bin"
```
Comment (jankeromnes) [inline full/nvm-lazy.sh:24]: Why do this here, as opposed to where Yarn is being installed?

**Review by jankeromnes — COMMENTED** [review]
Looks cool, thanks!

A few comments in passing.

File: .circleci/config.yml (hunk 1 above)
Comment (aledbf) [reply .circleci/config.yml:6]: because `pre-commit` package is being included since 20.10

File: CHANGELOG.md (hunk 2 above)
Comment (aledbf) [reply CHANGELOG.md:5]: fixed

File: full/nvm-lazy.sh (hunk 4 above)
Comment (aledbf) [reply full/nvm-lazy.sh:24]: fixed

File: full/nvm-lazy.sh (hunk 3 above)
Comment (aledbf) [reply full/nvm-lazy.sh]: fixed

**Review by csweichel — APPROVED** [review]
Makes sense to me.

File: full/nvm-lazy.sh [hunk 5]
```diff
@@ -1,18 +1,36 @@
...
+lazy_load_nvm() {
+    unset -f node npm npx yarn
+    export NVM_DIR="$HOME/.nvm"
+    # shellcheck disable=SC1091,SC1090
+    [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh" # This loads nvm
+    # shellcheck disable=SC1091,SC1090
+    [ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
+}
 
-declare -a NODE_GLOBALS=(`find ~/.nvm/versions/node -maxdepth 3 -type l -wholename '*/bin/*' | xargs -n1 basename | sort | uniq`)
```
Comment (jankeromnes) [inline full/nvm-lazy.sh:6]: FYI, if I run `find ~/.nvm/versions/node -maxdepth 3 -type l -wholename '*/bin/*' | xargs -n1 basename | sort | uniq` in a regular Gitpod workspace, I find all these:

```
node-gyp
npm
npx
tsc
tsserver
yarn
yarnpkg
```

Not necessary a deal-breaker, but why stop searching for binaries under `~/.nvm/versions/node`?

**Review by jankeromnes — APPROVED** [review]
Thanks for these improvements! Two more small things then go for launch. 🚀

File: full/nvm-lazy.sh (hunk 5 above)
Comment (aledbf) [reply full/nvm-lazy.sh:6]: @jankeromnes good point. Please check the last version.

What we really need is to lazy load nvm itself, not any of the commands.

File: SECURITY.md [hunk 6]
```diff
@@ -2,4 +2,4 @@
 
 We appreciate your efforts to responsibly disclose your findings, and will make every effort to acknowledge your contributions.
 
-To report a security issue please visit https://gitpod.io/security
+To report a security issue please visit [https://gitpod.io/security](https://gitpod.io/security)
```
[the remark this answers is not in the document]
Comment (aledbf) [reply SECURITY.md]: fixed

**Comment by aledbf** [comment]
Example of errors during validation of the changes
Screenshot from 2021-06-07 19-38-13

**Pull request merged**